testcases/kernel/syscalls/epoll2/man/epoll.4
.\"
.\"  epoll by Davide Libenzi ( efficient event notification retrieval )
.\"  Copyright (C) 2003  Davide Libenzi
.\"
.\"  This program is free software; you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License as published by
.\"  the Free Software Foundation; either version 2 of the License, or
.\"  (at your option) any later version.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program; if not, write to the Free Software
.\"  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.\"  Davide Libenzi <davidel@xmailserver.org>
.\"
.\"
.TH EPOLL 4 "23 October 2002" Linux "Linux Programmer's Manual"
.SH NAME
epoll \- I/O event notification facility
.SH SYNOPSIS
.B #include <sys/epoll.h>
.SH DESCRIPTION
.B epoll
is a variant of$
.BR poll (2)
that can be used either as Edge or Level Triggered interface and scales
well to large numbers of watched fds. Three system calls are provided to
set up and control an
.B epoll
set:$
.BR epoll_create (2) ,
.BR epoll_ctl (2) ,
.BR epoll_wait (2) .

An
.B epoll
set is connected to a file descriptor created by
.BR epoll_create (2) .
Interest for certain file descriptors is then registered via$
.BR epoll_ctl(2) .
Finally, the actual wait is started by$
.BR epoll_wait (2) .

.SH NOTES
The
.B epoll
event distribution interface is able to behave both as Edge Triggered
( ET ) and Level Triggered ( LT ). The difference between ET and LT
event distribution mechanism can be described as follows. Suppose that
this scenario happens :
.TP
.B 1
The file descriptor that represent the read side of a pipe (
.B RFD
) is added inside the
.B epoll
device.
.TP
.B 2
Pipe writer writes 2Kb of data on the write side of the pipe.
.TP
.B 3
A call to
.BR epoll_wait (2)
is done that will return
.B RFD
as ready file descriptor.
.TP
.B 4
The pipe reader reads 1Kb of data from
.BR RFD .
.TP
.B 5
A call to
.BR epoll_wait (2)
is done.
.PP

If the
.B RFD
file descriptor has been added to the
.B epoll
interface using the
.B EPOLLET
flag, the call to
.BR epoll_wait (2)
done in step
.B 5
will probably hang because of the available data still present in the file
input buffers and the remote peer might be expecting a response based on the
data it already sent. The reason for this is that Edge Triggered event
distribution delivers events only when events happens on the monitored file.
So, in step
.B 5
the caller might end up waiting for some data that is already present inside
the input buffer. In the above example, an event on
.B RFD
will be generated because of the write done in
.B 2
, and the event is consumed in
.BR 3 .
Since the read operation done in
.B 4
does not consume the whole buffer data, the call to
.BR epoll_wait (2)
done in step
.B 5
might lock indefinitely. The
.B epoll
interface, when used with the
.B EPOLLET
flag ( Edge Triggered )
should use non-blocking file descriptors to avoid having a blocking
read or write starve the task that is handling multiple file descriptors.
The suggested way to use
.B epoll
as an Edge Triggered (
.B EPOLLET
) interface is below, and possible pitfalls to avoid follow.
.SR
.TP$
.B i
with non-blocking file descriptors
.TP$
.B ii
by going to wait for an event only after
.BR read (2)
or$
.BR write (2)
return EAGAIN
.SE
.PP
On the contrary, when used as a Level Triggered interface,
.B epoll
is by all means a faster
.BR poll (2),
and can be used wherever the latter is used since it shares the
same semantics. Since even with the Edge Triggered
.B epoll
multiple events can be generated up on receival of multiple chunks of data,
the caller has the option to specify the
.B EPOLLONESHOT
flag, to tell
.B epoll
to disable the associated file descriptor after the receival of an event with
.BR epoll_wait (2).
When the
.B EPOLLONESHOT
flag is specified, it is caller responsibility to rearm the file descriptor using
.BR epoll_ctl (2)
with
.BR EPOLL_CTL_MOD.

.SH EXAMPLE FOR SUGGESTED USAGE

While the usage of
.B epoll
when employed like a Level Triggered interface does have the same
semantics of
.BR poll (2),
an Edge Triggered usage requires more clarifiction to avoid stalls
in the application event loop. In this example, listener is a
non-blocking socket on which
.BR listen (2)
has been called. The function do_use_fd() uses the new ready
file descriptor until EAGAIN is returned by either
.BR read (2)
or
.BR write (2).
An event driven state machine application should, after having received
EAGAIN, record its current state so that at the next call to do_use_fd()
it will continue to
.BR read (2)
or
.BR write (2)
from where it stopped before. $

.nf
struct epoll_event ev, *events;

for(;;) {
    nfds = epoll_wait(kdpfd, events, maxevents, -1);

    for(n = 0; n < nfds; ++n) {
        if(events[n].data.fd == listener) {
            client = accept(listener, (struct sockaddr *) &local,
                            &addrlen);
            if(client < 0){
                perror("accept");
                continue;
            }
            setnonblocking(client);
            ev.events = EPOLLIN | EPOLLET;
            ev.data.fd = client;
            if (epoll_ctl(kdpfd, EPOLL_CTL_ADD, client, &ev) < 0) {
                fprintf(stderr, "epoll set insertion error: fd=%d\n",
                        client);
                return -1;
            }
        }
        else
            do_use_fd(events[n].data.fd);
    }
}
.fi

When used as an Edge triggered interface, for performance reasons, it is
possible to add the file descriptor inside the epoll interface (
.B EPOLL_CTL_ADD
) once by specifying (
.BR EPOLLIN | EPOLLOUT
). This allows you to avoid
continuously switching between
.B EPOLLIN
and
.B EPOLLOUT
calling
.BR epoll_ctl (2)
with
.BR EPOLL_CTL_MOD .

.SH QUESTIONS AND ANSWERS (from linux-kernel)

.SR
.TP$
.B Q1$
What happens if you add the same fd to an epoll_set twice?
.TP
.B A1$
You will probably get EEXIST. However, it is possible that two
threads may add the same fd twice. This is a harmless condition.
.TP
.B Q2$
Can two
.B epoll
sets wait for the same fd? If so, are events reported
to both
.B epoll
sets fds?
.TP
.B A2
Yes. However, it is not recommended. Yes it would be reported to both.
.TP
.B Q3
Is the
.B epoll
fd itself poll/epoll/selectable?
.TP
.B A3
Yes.
.TP
.B Q4$
What happens if the
.B epoll
fd is put into its own fd set?
.TP
.B A4
It will fail. However, you can add an
.B epoll
fd inside another epoll fd set.$
.TP
.B Q5
Can I send the
.B epoll
fd over a unix-socket to another process?
.TP
.B A5
No.
.TP
.B Q6
Will the close of an fd cause it to be removed from all
.B epoll
sets automatically?
.TP
.B A6
Yes.
.TP
.B Q7$
If more than one event comes in between
.BR epoll_wait (2)
calls, are they combined or reported separately?
.TP
.B A7
They will be combined.
.TP
.B Q8
Does an operation on an fd affect the already collected but not yet reported
events?
.TP
.B A8
You can do two operations on an existing fd. Remove would be meaningless for
this case. Modify will re-read available I/O.
.TP
.B Q9
Do I need to continuously read/write an fd until EAGAIN when using the
.B EPOLLET
flag ( Edge Triggered behaviour ) ?
.TP
.B A9
No you don't. Receiving an event from
.BR epoll_wait (2)
should suggest to you that such file descriptor is ready for the requested I/O
operation. You have simply to consider it ready until you will receive the
next EAGAIN. When and how you will use such file descriptor is entirely up
to you. Also, the condition that the read/write I/O space is exhausted can
be detected by checking the amount of data read/write from/to the target
file descriptor. For example, if you call
.BR read (2)
by asking to read a certain amount of data and
.BR read (2)
returns a lower number of bytes, you can be sure to have exhausted the read
I/O space for such file descriptor. Same is valid when writing using the
.BR write (2)
function.
.SE

.SH POSSIBLE PITFALLS AND WAYS TO AVOID THEM
.SR
.TP
.B o Starvation ( Edge Triggered )
.PP
If there is a large amount of I/O space, it is possible that by trying to drain
it the other files will not get processed causing starvation. This
is not specific to
.BR epoll .
.PP
.PP
The solution is to maintain a ready list and mark the file descriptor as ready
in its associated data structure, thereby allowing the application to
remember which files need to be processed but still round robin amongst
all the ready files. This also supports ignoring subsequent events you
receive for fd's that are already ready.
.PP

.TP
.B o If using an event cache... $
.PP
If you use an event cache or store all the fd's returned from
.BR epoll_wait (2),
then make sure to provide a way to mark its closure dynamically (ie- caused by$
a previous event's processing). Suppose you receive 100 events from$
.BR epoll_wait (2),
and in eventi #47 a condition causes event #13 to be closed.$
If you remove the structure and close() the fd for event #13, then your$
event cache might still say there are events waiting for that fd causing$
confusion.
.PP
.PP$
One solution for this is to call, during the processing of event 47,
.BR epoll_ctl ( EPOLL_CTL_DEL )
to delete fd 13 and close(), then mark its associated
data structure as removed and link it to a cleanup list. If you find another
event for fd 13 in your batch processing, you will discover the fd had been
previously removed and there will be no confusion.
.PP

.SE
.SH CONFORMING TO
.BR epoll (4)
is a new API introduced in Linux kernel
.BR 2.5.44 .
Its interface should be finalized in Linux kernel
.BR 2.5.66 .
.SH "SEE ALSO"
.BR epoll_create (2)
.BR epoll_ctl (2)
.BR epoll_wait (2)



ChangeLog
LTP-20090228

1) Log Message:
We can just add the files related to LSM, to known failure list. We already check for their return value, if not EINVAL report test failure or else skip. Added the nfsd files to the list. Signed-Off-By: Kamalesh Babulal <kamalesh@linux.vnet.ibm.com>.

Modified File(s):
ltp/testcases/kernel/fs/proc/proc01.c

2) Log Message:
James Morris <jmorris@namei.org> reported: I'm trying to run the LTP SELinux tests using the latest CVS version of LTP and current Fedora development, and get the following policy compilation error: Is this likely to be fixed soon, and/or any suggestions for a workaround?
"Christopher J. PeBenito" <cpebenito@tresys.com> replied: It won't compile with the current trunk refpolicy, since the current release was a major, API breaking change.  I'll try to get a patch out shortly. I updated the policy since its fairly old, though I didn't convert its raw rules over to use interfaces.  However this didn't completely fix it, as there is usage of a "unconfined_runs_test()", which isn't in the upstream refpolicy nor the fedora policy, as far as I can see.  One of the updates includes use of sysadm_entry_spec_domtrans_to(), which is in the upstream refpolicy, but doesn't seem to have made its way downstream to the fedora policy.  I have attached my work so someone familiar with the LTP test cases can use it to complete the fix. Signed-Off-By: "Christopher J. PeBenito" <cpebenito@tresys.com>.

Modified File(s):
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_capable_file.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_capable_net.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_dyntrace.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_dyntrans.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_entrypoint.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_execshare.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_exectrace.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_execute_no_trans.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_fdreceive.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_file.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_global.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_inherit.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_ioctl.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_ipc.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_link.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_mkdir.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_open.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_ptrace.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_relabel.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_rename.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_setattr.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_setnice.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_sigkill.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_sysctl.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_task_create.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_task_getpgid.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_task_getsched.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_task_getsid.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_task_setpgid.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_task_setsched.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_transition.te
ltp/testcases/kernel/security/selinux-testsuite/refpolicy/test_wait.te

3) Log Message:
I send attached the v4l-test 0.5 for LTP. Changes: Test cases added for VIDIOC_QUERYSTD, VIDIOC_G_FREQUENCY and VIDIOC_G_TUNER. Signed-off-by: Márton Németh <nm127@freemail.hu>.

Modified Files:
ltp/testcases/kernel/device-drivers/v4l/user_space/Makefile
ltp/testcases/kernel/device-drivers/v4l/user_space/README
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_STD.c
ltp/testcases/kernel/device-drivers/v4l/user_space/v4l2_test.c
ltp/testcases/kernel/device-drivers/v4l/user_space/doc/index.html
Added Files:
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_FREQUENCY.c
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_FREQUENCY.h
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_QUERYSTD.c
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_QUERYSTD.h
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_TUNER.c
ltp/testcases/kernel/device-drivers/v4l/user_space/test_VIDIOC_TUNER.h
ltp/testcases/kernel/device-drivers/v4l/user_space/v4l2_validator.c
ltp/testcases/kernel/device-drivers/v4l/user_space/v4l2_validator.h

4) Log Message:
Fix a build failure of tcore on non-x86 archs: I report a build failure and a patch for it. If your system is not either i386 or x86_64, making in ltp-full-20090131 fails as follows: It's because wrong position of TST_TOTAL definition. Signed-off-by: Seiichi Ikarashi <s.ikarashi@jp.fujitsu.com>.

Modified File(s):
ltp/testcases/misc/tcore_patch_test_suites/tcore.c

5) Log Message:
Fix build failures of some new syscall tests on non-x86 archs: I report build failures of some new syscall tests introduced in ltp-full-20090131. Failures happen on non-x86 arch systems.I made a patch for it; stopping #error macro and including linux_syscall_numbers.h instead. Signed-off-by: Seiichi Ikarashi <s.ikarashi@jp.fujitsu.com>.

Modified File(s):
ltp/testcases/kernel/include/i386.in
ltp/testcases/kernel/include/regen.sh
ltp/testcases/kernel/include/x86_64.in
ltp/testcases/kernel/syscalls/dup3/dup3_01.c
ltp/testcases/kernel/syscalls/epoll_create2/epoll_create2_01.c

testcases/kernel/syscalls/libevent/epoll.c
/*
 * Copyright 2000-2003 Niels Provos <provos@citi.umich.edu>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdint.h>
#include <sys/types.h>
#include <sys/resource.h>
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#else
#include <sys/_time.h>
#endif
#include <sys/queue.h>
#include <sys/epoll.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <err.h>

#ifdef USE_LOG
#include "log.h"
#else
#define LOG_DBG(x)
#define log_error	warn
#endif

#include "event.h"
#include "evsignal.h"

extern struct event_list eventqueue;

extern volatile sig_atomic_t evsignal_caught;

/* due to limitations in the epoll interface, we need to keep track of
 * all file descriptors outself.
 */
struct evepoll {
	struct event *evread;
	struct event *evwrite;
};

struct epollop {
	struct evepoll *fds;
	int nfds;
	struct epoll_event *events;
	int nevents;
	int epfd;
	sigset_t evsigmask;
} epollop;

void *epoll_init(void);
int epoll_add(void *, struct event *);
int epoll_del(void *, struct event *);
int epoll_recalc(void *, int);
int epoll_dispatch(void *, struct timeval *);

struct eventop epollops = {
	"epoll",
	epoll_init,
	epoll_add,
	epoll_del,
	epoll_recalc,
	epoll_dispatch
};

#define NEVENT	32000

void *epoll_init(void)
{
	int epfd, nfiles = NEVENT;
	struct rlimit rl;

	/* Disable epollueue when this environment variable is set */
	if (getenv("EVENT_NOEPOLL"))
		return (NULL);

	memset(&epollop, 0, sizeof(epollop));

	if (getrlimit(RLIMIT_NOFILE, &rl) == 0 && rl.rlim_cur != RLIM_INFINITY)
		nfiles = rl.rlim_cur;

	/* Initalize the kernel queue */

	if ((epfd = epoll_create(nfiles)) == -1) {
		log_error("epoll_create");
		return (NULL);
	}

	epollop.epfd = epfd;

	/* Initalize fields */
	epollop.events = malloc(nfiles * sizeof(struct epoll_event));
	if (epollop.events == NULL)
		return (NULL);
	epollop.nevents = nfiles;

	epollop.fds = calloc(nfiles, sizeof(struct evepoll));
	if (epollop.fds == NULL) {
		free(epollop.events);
		return (NULL);
	}
	epollop.nfds = nfiles;

	evsignal_init(&epollop.evsigmask);

	return (&epollop);
}

int epoll_recalc(void *arg, int max)
{
	struct epollop *epollop = arg;

	if (max > epollop->nfds) {
		struct evepoll *fds;
		int nfds;

		nfds = epollop->nfds;
		while (nfds < max)
			nfds <<= 1;

		fds = realloc(epollop->fds, nfds * sizeof(struct evepoll));
		if (fds == NULL) {
			log_error("realloc");
			return (-1);
		}
		epollop->fds = fds;
		memset(fds + epollop->nfds, 0,
		       (nfds - epollop->nfds) * sizeof(struct evepoll));
		epollop->nfds = nfds;
	}

	return (evsignal_recalc(&epollop->evsigmask));
}

int epoll_dispatch(void *arg, struct timeval *tv)
{
	struct epollop *epollop = arg;
	struct epoll_event *events = epollop->events;
	struct evepoll *evep;
	int i, res, timeout;

	if (evsignal_deliver(&epollop->evsigmask) == -1)
		return (-1);

	timeout = tv->tv_sec * 1000 + tv->tv_usec / 1000;
	res = epoll_wait(epollop->epfd, events, epollop->nevents, timeout);

	if (evsignal_recalc(&epollop->evsigmask) == -1)
		return (-1);

	if (res == -1) {
		if (errno != EINTR) {
			log_error("epoll_wait");
			return (-1);
		}

		evsignal_process();
		return (0);
	} else if (evsignal_caught)
		evsignal_process();

	LOG_DBG((LOG_MISC, 80, "%s: epoll_wait reports %d", __func__, res));

	for (i = 0; i < res; i++) {
		int which = 0;
		int what = events[i].events;
		struct event *evread = NULL, *evwrite = NULL;

		evep = (struct evepoll *)events[i].data.ptr;

		if (what & EPOLLHUP)
			what |= EPOLLIN | EPOLLOUT;
		else if (what & EPOLLERR)
			what |= EPOLLIN | EPOLLOUT;

		if (what & EPOLLIN) {
			evread = evep->evread;
			which |= EV_READ;
		}

		if (what & EPOLLOUT) {
			evwrite = evep->evwrite;
			which |= EV_WRITE;
		}

		if (!which)
			continue;

		if (evread != NULL && !(evread->ev_events & EV_PERSIST))
			event_del(evread);
		if (evwrite != NULL && evwrite != evread &&
		    !(evwrite->ev_events & EV_PERSIST))
			event_del(evwrite);

		if (evread != NULL)
			event_active(evread, EV_READ, 1);
		if (evwrite != NULL)
			event_active(evwrite, EV_WRITE, 1);
	}

	return (0);
}

int epoll_add(void *arg, struct event *ev)
{
	struct epollop *epollop = arg;
	struct epoll_event epev;
	struct evepoll *evep;
	int fd, op, events;

	if (ev->ev_events & EV_SIGNAL)
		return (evsignal_add(&epollop->evsigmask, ev));

	fd = ev->ev_fd;
	if (fd >= epollop->nfds) {
		/* Extent the file descriptor array as necessary */
		if (epoll_recalc(epollop, fd) == -1)
			return (-1);
	}
	evep = &epollop->fds[fd];
	op = EPOLL_CTL_ADD;
	events = 0;
	if (evep->evread != NULL) {
		events |= EPOLLIN;
		op = EPOLL_CTL_MOD;
	}
	if (evep->evwrite != NULL) {
		events |= EPOLLOUT;
		op = EPOLL_CTL_MOD;
	}

	if (ev->ev_events & EV_READ)
		events |= EPOLLIN;
	if (ev->ev_events & EV_WRITE)
		events |= EPOLLOUT;

	epev.data.ptr = evep;
	epev.events = events;
	if (epoll_ctl(epollop->epfd, op, ev->ev_fd, &epev) == -1)
		return (-1);

	/* Update events responsible */
	if (ev->ev_events & EV_READ)
		evep->evread = ev;
	if (ev->ev_events & EV_WRITE)
		evep->evwrite = ev;

	return (0);
}

int epoll_del(void *arg, struct event *ev)
{
	struct epollop *epollop = arg;
	struct epoll_event epev;
	struct evepoll *evep;
	int fd, events, op;
	int needwritedelete = 1, needreaddelete = 1;

	if (ev->ev_events & EV_SIGNAL)
		return (evsignal_del(&epollop->evsigmask, ev));

	fd = ev->ev_fd;
	if (fd >= epollop->nfds)
		return (0);
	evep = &epollop->fds[fd];

	op = EPOLL_CTL_DEL;
	events = 0;

	if (ev->ev_events & EV_READ)
		events |= EPOLLIN;
	if (ev->ev_events & EV_WRITE)
		events |= EPOLLOUT;

	if ((events & (EPOLLIN | EPOLLOUT)) != (EPOLLIN | EPOLLOUT)) {
		if ((events & EPOLLIN) && evep->evwrite != NULL) {
			needwritedelete = 0;
			events = EPOLLOUT;
			op = EPOLL_CTL_MOD;
		} else if ((events & EPOLLOUT) && evep->evread != NULL) {
			needreaddelete = 0;
			events = EPOLLIN;
			op = EPOLL_CTL_MOD;
		}
	}

	epev.events = events;
	epev.data.ptr = evep;

	if (epoll_ctl(epollop->epfd, op, fd, &epev) == -1)
		return (-1);

	if (needreaddelete)
		evep->evread = NULL;
	if (needwritedelete)
		evep->evwrite = NULL;

	return (0);
}
